clang/include/clang/CIR/Dialect/IR/CIRTypes.td

```diff
@@ -94,6 +95,85 @@ def SInt16 : SInt<16>;
 def SInt32 : SInt<32>;
 def SInt64 : SInt<64>;
 
+//===----------------------------------------------------------------------===//
+// FloatType
+//===----------------------------------------------------------------------===//
+
+def CIR_FloatType : CIR_Type<"Float", "float",
+    [DeclareTypeInterfaceMethods<DataLayoutTypeInterface>]> {
+  let summary = "CIR floating-point type";
+  let description = [{
+    CIR type that represents C/C++ floating-point types, including `float`, \
+    `double`, and `long double`.
+  }];
+  let parameters = (ins "::mlir::cir::FloatType::Format":$format);
+
+  let builders = [
+    TypeBuilder<(ins CArg<"::mlir::cir::FloatType::Format">:$format)>
+  ];
+
+  let hasCustomAssemblyFormat = 1;
+  let genVerifyDecl = 1;
+
+  let extraClassDeclaration = [{
+    // Floating-point format.
+    enum Format : uint32_t {
+      FF_Binary16,   // IEEE-754 binary16 format
+      FF_Binary32,   // IEEE-754 binary32 format
+      FF_Binary64,   // IEEE-754 binary64 format
+      FF_Binary64E,  // IEEE-754 binary64-extended format
+      FF_Binary128,  // IEEE-754 binary128 format
+      FF_BFloat16,   // Brain floating format
+    };
```

Comment:
This covers six of the seven [LLVM IR floating-point types](https://llvm.org/docs/LangRef.html#floating-point-types).  I suggest that we just go ahead and add the seventh type now, since we will need it eventually.  It is `ppc_fp128`, commonly known as "double-double", and is used to represent `long double` on most Linux Power systems.

Reply:
I have added `FF_PpcFp128` format to the `cir.float` type. However I met a problem when trying to lower it to LLVMIR. See [my latest comment](https://github.com/llvm/clangir/pull/385#issuecomment-1892290023).